### PR TITLE
Validate distgit config files earlier

### PR DESCRIPTION
**Pull request by tkdchen**
* CLOUDBLD-1989

This patch validates fetch-artifacts-*.yaml and content_sets.yml in the
pre_check_user_settings plugin.

Signed-off-by: Chenxiong Qi <qcxhome@gmail.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [n/a] JSON/YAML configuration changes are updated in the relevant schema
- [n/a] Changes to metadata also update the documentation for the metadata
- [n/a] Pull request has a link to an osbs-docs PR for user documentation updates
- [n/a] New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/constants.py [hunk 1]
```diff
@@ -189,3 +192,21 @@
 
 # location for build info directory in the image
 IMAGE_BUILD_INFO_DIR = '/root/buildinfo/'
+
+
+DistgitConfigFile = namedtuple('DistgitConfigFile', ['filename', 'schema'])
```
Comment (MartinBasti) [inline atomic_reactor/constants.py:197]: `DistGit` is implementation detail, let's call it more generic `OSBSConfigFile`, `GitConfigFile` or `UserConfigFile`, or something like that.

Comment (MartinBasti) [reply atomic_reactor/constants.py:197]: or `InRepoConfigFile`

File: atomic_reactor/plugins/pre_resolve_composes.py [hunk 2]
```diff
@@ -211,11 +208,15 @@ def read_configs(self):
...
-        workdir = self.workflow.source.get_build_file_path()[1]
-        file_path = os.path.join(workdir, REPO_CONTENT_SETS_CONFIG)
-        pulp_data = None
-        if os.path.exists(file_path):
-            pulp_data = read_yaml_from_file_path(file_path, 'schemas/content_sets.json') or {}
+        # workdir = self.workflow.source.get_build_file_path()[1]
+        # file_path = os.path.join(workdir, REPO_CONTENT_SETS_CONFIG)
+        # pulp_data = None
+        # if os.path.exists(file_path):
+        #     pulp_data = read_yaml_from_file_path(file_path, 'schemas/content_sets.json') or {}
```
Comment (MartinBasti) [inline atomic_reactor/plugins/pre_resolve_composes.py:215]: :eyes:

File: atomic_reactor/plugins/pre_resolve_composes.py [hunk 3]
```diff
@@ -211,11 +208,15 @@ def read_configs(self):
...
+
+        from atomic_reactor import util
+        pulp_data = util.read_content_sets(self.workflow)
+        # breakpoint()
```
Comment (MartinBasti) [inline atomic_reactor/plugins/pre_resolve_composes.py:219]: :eyes:

File: atomic_reactor/plugins/pre_resolve_composes.py [hunk 4]
```diff
@@ -211,11 +208,15 @@ def read_configs(self):
...
-        pulp_data = None
-        if os.path.exists(file_path):
-            pulp_data = read_yaml_from_file_path(file_path, 'schemas/content_sets.json') or {}
+        # workdir = self.workflow.source.get_build_file_path()[1]
+        # file_path = os.path.join(workdir, REPO_CONTENT_SETS_CONFIG)
+        # pulp_data = None
+        # if os.path.exists(file_path):
+        #     pulp_data = read_yaml_from_file_path(file_path, 'schemas/content_sets.json') or {}
+
+        from atomic_reactor import util
```
Comment (MartinBasti) [inline atomic_reactor/plugins/pre_resolve_composes.py:217]: Can we have this import at the top?

File: tests/plugins/test_resolve_composes.py [hunk 5]
```diff
@@ -1561,7 +1561,7 @@ def test_canceling_compose_when_timeout_of_waiting_for_the_compose(
         # Ensure ODCSClient.wait_for_compose raises timeout error
         (flexmock(time)
          .should_receive('time')
-         .and_return(2 + ODCSClient.DEFAULT_WAIT_TIMEOUT, 1)
```
Comment (MartinBasti) [inline tests/plugins/test_resolve_composes.py:1564]: Why this change?  If it's unrelated fix it should be at least in separate commit

Comment (tkdchen) [reply tests/plugins/test_resolve_composes.py:1564]: It was detected during running tests with the new code changes. I'm actually confused why this change is needed. But, it just works now. :)

**Review by MartinBasti — APPROVED** [review]

**Review by ben-alkov — APPROVED** [review]

**Review by lkolacek — APPROVED** [review]

File: atomic_reactor/constants.py [hunk 6]
```diff
@@ -189,3 +192,21 @@
 
 # location for build info directory in the image
 IMAGE_BUILD_INFO_DIR = '/root/buildinfo/'
+
+
+UserConfigFile = namedtuple('UserConfigFile', ['filename', 'schema'])
+
+USER_CONFIG_FILES = {
```
Comment (rcerven) [inline atomic_reactor/constants.py:199]: why not have just simple dict 
{REPO_FETCH_ARTIFACTS_URL: 'schemas/fetch-artifacts-url.json', .... } 
?

having the same value in key and then again in filename seems a waste

File: atomic_reactor/constants.py [hunk 7]
```diff
@@ -161,6 +162,8 @@
 
 REPO_CONTAINER_CONFIG = 'container.yaml'
 REPO_CONTENT_SETS_CONFIG = 'content_sets.yml'
+REPO_FETCH_ARTIFACTS_URL = 'fetch-artifacts-url.yaml'
```
Comment (rcerven) [inline atomic_reactor/constants.py:165]: fetch artifacts plugin had originally 

NVR_REQUESTS_FILENAME = 'fetch-artifacts-koji.yaml'
URL_REQUESTS_FILENAME = 'fetch-artifacts-url.yaml'

why are we renaming configuration files???   from 'fetch-artifacts-nvr.yaml' to 'fetch-artifacts-koji.yaml'

Comment (rcerven) [inline atomic_reactor/constants.py:165]: fetch artifacts plugin had originally

NVR_REQUESTS_FILENAME = 'fetch-artifacts-koji.yaml'
URL_REQUESTS_FILENAME = 'fetch-artifacts-url.yaml'

why are we renaming configuration files??? from 'fetch-artifacts-koji.yaml' to 'fetch-artifacts-nvr.yaml'

File: atomic_reactor/constants.py (hunk 6 above)
Comment (rcerven) [inline atomic_reactor/constants.py:199]: why not have just simple dict
{REPO_FETCH_ARTIFACTS_URL: 'schemas/fetch-artifacts-url.json', .... }
?

having the same value in key and then again in filename seems a waste

File: atomic_reactor/constants.py (hunk 7 above)
Comment (tkdchen) [reply atomic_reactor/constants.py:164]: @rcerven Good catch! Sorry for this mistake. It is fixed in latest updated commit.

File: atomic_reactor/constants.py (hunk 6 above)
Comment (tkdchen) [reply atomic_reactor/constants.py:195]: Updated. PTAL.

**Review by MartinBasti — APPROVED** [review]

**Review by mkosiarc — APPROVED** [review]

File: atomic_reactor/plugins/pre_resolve_composes.py [hunk 8]
```diff
@@ -211,11 +209,7 @@ def read_configs(self):
         if not data and not self.all_compose_ids:
             raise SkipResolveComposesPlugin('"compose" config not set and compose_ids not given')
 
-        workdir = self.workflow.source.get_build_file_path()[1]
-        file_path = os.path.join(workdir, REPO_CONTENT_SETS_CONFIG)
-        pulp_data = None
-        if os.path.exists(file_path):
-            pulp_data = read_yaml_from_file_path(file_path, 'schemas/content_sets.json') or {}
+        pulp_data = util.read_content_sets(self.workflow)
```
Comment (rcerven) [inline atomic_reactor/plugins/pre_resolve_composes.py:212]: missing : or {}
OR you could also return those default values []/{} in those util reading methods

**Review by rcerven — CHANGES_REQUESTED** [review]

File: atomic_reactor/plugins/pre_resolve_composes.py (hunk 8 above)
Comment (rcerven) [reply atomic_reactor/plugins/pre_resolve_composes.py:212]: actually it doesn't matter in this case that much, but still

it would be good to provide those default values in reading methods, and then you can remove those fallbacks to []/{}

**Review by rcerven — APPROVED** [review]

**Comment by rcerven** [comment]
rn updated: Validate distgit config files earlier
